src/coprocessor/codec/mysql/decimal.rs

```diff
@@ -528,22 +528,45 @@ fn do_div_mod(
     do_mod: bool,
 ) -> Option<Res<Decimal>> {
     let r_frac_cnt = word_cnt!(rhs.frac_cnt) * DIGITS_PER_WORD;
-    let (r_idx, mut r_prec) = rhs.remove_leading_zeroes();
-    r_prec += r_frac_cnt;
-    if r_prec == 0 {
-        return None;
-    }
+    let (r_idx, r_prec) = {
+        let mut prec = rhs.int_cnt + r_frac_cnt;
+        let mut i = ((prec + DIGITS_PER_WORD - 1) % DIGITS_PER_WORD) + 1;
+        let mut word_idx = 0;
+        while prec > 0 && rhs.word_buf[word_idx] == 0 {
+            prec -= i;
+            i = DIGITS_PER_WORD;
+            word_idx += 1;
+        }
+        if prec == 0 {
+            return None;
+        } else {
+            prec -= count_leading_zeroes((prec - 1) % DIGITS_PER_WORD, rhs.word_buf[word_idx])
+        }
+        (word_idx, prec)
+    };
 
     let l_frac_cnt = word_cnt!(lhs.frac_cnt) * DIGITS_PER_WORD;
-    let (l_idx, mut l_prec) = lhs.remove_leading_zeroes();
-    l_prec += l_frac_cnt;
-    if l_prec == 0 {
-        lhs.reset_to_zero();
-        return Some(Res::Ok(lhs));
-    }
+    let (l_idx, l_prec) = {
+        let mut prec = lhs.int_cnt + l_frac_cnt;
+        let mut i = ((prec + DIGITS_PER_WORD - 1) % DIGITS_PER_WORD) + 1;
+        let mut word_idx = 0;
+        while prec > 0 && lhs.word_buf[word_idx] == 0 {
+            prec -= i;
+            i = DIGITS_PER_WORD;
+            word_idx += 1;
+        }
+        if prec == 0 {
+            lhs.reset_to_zero();
+            return Some(Res::Ok(lhs));
+        } else {
+            prec -= count_leading_zeroes((prec - 1) % DIGITS_PER_WORD, lhs.word_buf[word_idx])
+        }
+        (word_idx, prec)
+    };
```

Comment:
Could you define L549~525 to an independent function?

Reply:
I don't get the semantic of the function: the code is a little bit tricky but very similar to remove_leading_zeros(), would we name the function something like remove_all_leading_zeros or calculate_idx_and_prec?

Reply:
Maybe you could transfer `lhs.int_cnt+l_frac_cnt` as a parameter for `remove_leading_zeroes` here?